File: Algo/Leetcode/242ValidAnagram.cpp
// sort. O(nlogn)
class Solution {
public:
    bool isAnagram(string s, string t) {
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());
        return s==t;
    }
};

//Use dict. O(n)
class Solution {
public:
    bool isAnagram(string s, string t) {
        vector<int> m(256, 0);
        for(auto c:s) m[c]+=1;
        for(auto c:t) m[c]-=1;
        for(auto c:m){
            if (c!=0) return false;
        }
        return true;
    }
};

File: Algo/Leetcode/242ValidAnagram.py
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        if len(s) < len(t):
            return self.isAnagram(t, s)

        d = {}
        for n in t:
            d[n] = d.get(n, 0) + 1

        for n in s:
            if d.get(n, 0) == 0:
                return False
            else:
                d[n] -= 1

        return True